### DISCUSSION TITLE
Develop a GLSP diagram extension

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-glsp/glsp#833] [Q&A]
Hi,
I try to develop an extension on theia blueprint, this extension is a glsp diagram. I have seen documents about glsp and theia-integration. But that demo is using glsp as the main program, in my application glsp is just an extension, I have some other extensions in my program. I am not very sure about the architecture of glsp as an extension. Should I develop one glsp-client as an extension and one glsp-server as another extension? or I create one extension which contains the client and server?
I use node-json-theia solution. any experienced person can give some advice, thank you very much

**Comment by planger** [comment]
Hi Pagina,

You are right, the `node-json-theia` project already contains a Theia application (`tasklist-browser-app`). This is to make it as easy as possible to get started.

Basically, `tasklist-theia` is the Theia extension and `tasklist-glsp` is the diagram implementation, that `tasklist-theia` integrates into a Theia. The app `tasklist-browser-app` is just an example Theia app, that includes `tasklist-theia`.

The server is independent from all of that and needs to run and expose a port alongside the Theia application.
In this context, the following parts of the documentation may help:

* https://www.eclipse.org/glsp/documentation/integrations/
* https://theia-ide.org/docs/authoring_extensions/
* https://theia-ide.org/docs/composing_applications/


If you want to use this diagram extension now in the context of Theia Blueprint, you have two options:
1. You publish the GLSP diagram packages `tasklist-theia` and its dependency `tasklist-glsp` to npmjs and add `tasklist-theia` to the dependencies in the Theia Blueprint app's [`package.json`](https://github.com/eclipse-theia/theia-blueprint/blob/master/applications/electron/package.json#L44). You'd still also need to start the server externally.
2. You directly move the sources of the GLSP diagram packages into Theia Blueprint and build it there directly. This is recommended if you want to co-develop your Theia applicatoin and your diagram editor and avoid a publishing step in between every time you change something.

For approach 2, you'd move `tasklist-glsp` and `tasklist-theia` into [`theia-extensions`](https://github.com/eclipse-theia/theia-blueprint/tree/master/theia-extensions) and add `tasklist-theia` to the dependencies in the Theia Blueprint app's [`package.json`](https://github.com/eclipse-theia/theia-blueprint/blob/master/applications/electron/package.json#L44).

There might be some additional configuration changes required (tsconfig), but roughly that should work.

**Reply by wss29** [reply]
Hi,
    Thank you very much for your advice. Now I download glsp-theia-integration and try to use workflow-theia as a demo extension to my project.
1. I move the sources of theia-integration and workflow-theia to theia-extensions folder
![image](https://user-images.githubusercontent.com/4537809/209420808-bb90b5a9-86c4-45dc-9095-c75ca70f3ae3.png)
2. I add follow resolutions to the root package.json
![image](https://user-images.githubusercontent.com/4537809/209420819-c63757f7-70b6-46a7-bc50-2ad2c0745499.png)
3.I add workflow-theia to Blueprint app's package.json
![image](https://user-images.githubusercontent.com/4537809/209420837-e7c08890-e743-4456-bb47-1b2d93e780ed.png)
4.I yarn and build, and everything is ok, But when I click example1.wf file,nothing happens, can not open the wf file
![image](https://user-images.githubusercontent.com/4537809/209421066-cee97d4e-a10b-4bc7-853a-cfd8fd2d4d5a.png)
5.I guess the server is not starting, so I check the jar file, It exists. 
![image](https://user-images.githubusercontent.com/4537809/209420939-3cdafae6-8051-450f-b0a4-5cceb2e0ba73.png)
6. I check task manager and see the two java.exe run which is the same as normal as glsp-theia-integration. So the server should run, but I do not know why .wf can not open, have no idea
![image](https://user-images.githubusercontent.com/4537809/209420959-62aae3bb-9a19-4ab2-b67d-9b70e0bbca2e.png)

**Reply by planger** [reply]
Unfortunately, I can't tell what's wrong without further details. Can you please also provide the browser console output and the Theia backend logs?
